### DISCUSSION TITLE
How to add a value for a link property

### DISCUSSION THREAD

**Discussion by clmnin** [geldata/gel#2817] [Q&A] [answered]
There are mentions of [link-properties](https://www.edgedb.com/docs/edgeql/expressions/shapes#link-properties) but no docs on `INSERT`ing a link property.

Here is something I started working on, intuitively
```sql
START MIGRATION TO {
    module default {
        scalar type OrderNo extending sequence;
        type User {
            required property name -> str {
                constraint exclusive;
            }
            multi link friends -> User {
                property nickname -> str {
                    constraint max_len_value(200);
                }
            }
        }
    }
};
POPULATE MIGRATION;
COMMIT MIGRATION;

INSERT User {name := 'Alice'};
INSERT User {name := 'Bob'};
INSERT User {name := 'Cathy'};

UPDATE User FILTER .name = "Alice" SET { friends := (
    SELECT DETACHED User {
        @nickname := "Bob the builder"
    } FILTER .name = "Bob"
) };
```

But it did not work as expected. I get this error
> error: invalid reference to link property in top level shape

**Comment by vpetrovykh** [comment, marked as answer]
Oh! Good catch!

The error currently comes from the use of `DETACHED`. This is a tricky bit semantically. Even as a developer of EdgeDB I'm not certain whether the error here is what I would expect to happen or not.

However, as a general rule we would recommend that you avoid using an explicit `DETACHED` in favour of aliasing things in the WITH block. `DETACHED` can be harder to manage for non-trivial expressions.

So try something like this instead:
```
WITH U := User
UPDATE User FILTER .name = "Alice" SET { friends := (
    SELECT U {
        @nickname := "Bob the builder"
    } FILTER .name = "Bob"
) };
```

Just in case you missed it, check out our cheat sheets:
https://www.edgedb.com/docs/cheatsheet/update

There are a couple of examples of dealing with link properties in `UPDATE` that you may find useful.

**Reply by elprans** [reply]
The `DETACHED` thing here is a known issue.  Essentially, the compiler is unable to "see" through layers of expressions to attribute the link property to the correct link.  See #2038 for another example.  We're working on fixing this in general, meanwhile the workaround suggested by @vpetrovykh is good.
